Upgrade to Rails 5

kalamkar opened the pull request.

kalamkar commented on Gemfile:
```diff
@@ -17,15 +17,14 @@ gem 'pg'
-
-# for sending devise emails in background
-gem 'devise-async'
+gem 'devise' , '~> 4.1.0'
```
Removed `devise-async` with reference to the issue https://github.com/bigbinary/aapc7atlis/pull/3531

kalamkar commented on Gemfile:
```diff
@@ -17,15 +17,14 @@ gem 'pg'
+gem "sinatra", github: 'sinatra/sinatra'
+gem "rack-protection", github: 'sinatra/rack-protection'
 gem 'delayed_job_web', '>= 1.2.10'
 
```
Reference https://github.com/ejschmitt/delayed_job_web/issues/84

kalamkar commented on Gemfile:
```diff
@@ -48,6 +47,7 @@ gem 'simple_form'
 # admin framework
 gem 'activeadmin', git: 'https://github.com/activeadmin/activeadmin.git'
+gem 'inherited_resources', github: 'activeadmin/inherited_resources'
 
```
Reference https://github.com/activeadmin/activeadmin/issues/4177#issuecomment-231680674

kalamkar commented on Gemfile:
```diff
@@ -80,16 +80,13 @@ gem 'email_prefixer'
 # Attach comments to Active Record queries
-gem 'marginalia'
+gem 'marginalia', github: 'basecamp/marginalia'
 
```
Reference https://github.com/basecamp/marginalia/issues/58

New version with latest changes is not yet released

kalamkar commented on Gemfile:
```diff
@@ -80,16 +80,13 @@ gem 'email_prefixer'
 
-  # mutes assets pipeline log messages
-  gem 'quiet_assets'
-
```
It is now part of Rails 5 https://github.com/bigbinary/bbjekyll/issues/579

kalamkar commented on Gemfile:
```diff
@@ -108,4 +105,6 @@ group :test do
   gem 'simplecov', require: false
 
+  # for controller testing
+  gem 'rails-controller-testing'
```
As mentioned in the upgrading guide - 

`assigns` and `assert_template` have been extracted to the `rails-controller-testing` gem 

http://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html

neerajdotname replied:
Let's change the test to not use `assert_template`. This gem should be removed.

kalamkar replied:
@neerajdotname Just checked [this comment by DHH](https://github.com/rails/rails/issues/18950#issue-57746192). It makes sense to remove `assigns` and `assert_template`.

kalamkar commented on config/environments/test.rb:
```diff
@@ -13,8 +13,8 @@
-  config.static_cache_control = 'public, max-age=3600'
+  config.public_file_server.enabled  = true
+  config.public_file_server.headers  = { 'Cache-Control' => 'public, max-age=3600' }
 
```
Previous syntax is deprecated.

neerajdotname commented:
@kalamkar Thanks for the quick turn around. Appreciate it.

Tomorrow sometime can you run `rails new` on a new project and see if all the files that are created and the values that are there in those files are present in this PR. While upgrading Rails that is the biggest issue we have in wheel since we do not generate new `rails new` so it is best to double check that.

kalamkar commented on config/initializers/catch_json_parser_errors.rb:
```diff
@@ -1,3 +1,3 @@
 unless Rails.env.development?
-  Rails.application.config.middleware.insert_before ActionDispatch::ParamsParser, "CatchJsonParseErrors"
+  Rails.application.config.middleware.use CatchJsonParseErrors
 end
```
`ActionDispatch::ParamsParser` is not a part of Rails middleware.

Available middleware

```
> rails middleware

use Rack::Sendfile
use ActionDispatch::Static
use ActionDispatch::Executor
use ActiveSupport::Cache::Strategy::LocalCache::Middleware
use Rack::Timeout
use Rack::Runtime
use Rack::MethodOverride
use ActionDispatch::RequestId
use Rails::Rack::Logger
use ActionDispatch::ShowExceptions
use WebConsole::Middleware
use ActionDispatch::DebugExceptions
use ActionDispatch::RemoteIp
use ActionDispatch::Reloader
use ActionDispatch::Callbacks
use ActiveRecord::Migration::CheckPending
use ActionDispatch::Cookies
use ActionDispatch::Session::CookieStore
use ActionDispatch::Flash
use Rack::Head
use Rack::ConditionalGet
use Rack::ETag
use Warden::Manager
use Bullet::Rack
use Rack::Deflater
run Wheel::Application.routes
```

neerajdotname replied:
@kalamkar please remove this middleware `CatchJsonParseErrors`. It causes more problems than helping us.

kalamkar commented:
@neerajdotname 

> Tomorrow sometime can you run rails new on a new project and see if all the files that are created and the values that are there in those files are present in this PR

Yes, I will try to check it in this weekend.

> please remove this middleware CatchJsonParseErrors. It causes more problems than helping us.

Will remove it.

kalamkar commented:
@neerajdotname _a

> please remove this middleware CatchJsonParseErrors. It causes more problems than helping us.

Commit - https://github.com/bigbinary/wheel/pull/120/commits/24f780efdc329935a6fce879fd484562faadc4b0

1. Upgraded Devise gem version to 4.2.0
2. Added Rails 5 ActionController API support
3. Removed catch json parse error middleware
4. Fixed failing test case.
5. Modified deprecated Devise helper module

> Tomorrow sometime can you run rails new on a new project and see if all the files that are created and the values that are there in those files are present in this PR

Commit - https://github.com/bigbinary/wheel/pull/120/commits/0b03003de638369e6f552d25833f4f12f8c39dd4

1. Modified the previous Rails version files with Rails 5 file changes.
2. Added missing Rails 5 application files to the application.

Please review. Let me know if any setting is missing/not required after the changes.

neerajdotname commented on app/controllers/application_controller.rb:
```diff
@@ -2,34 +2,8 @@ class ApplicationController < ActionController::Base
   # Prevent CSRF attacks by raising an exception.
   # For APIs, you may want to use :null_session instead.
-  protect_from_forgery with: :exception
+  protect_from_forgery unless: -> { request.format.json? }
```
Is this a Rails 5 change?

kalamkar replied:
It was not default Rails 5 change. Removed it.

neerajdotname commented on app/controllers/api/v1/base_controller.rb:
```diff
@@ -1,10 +1,12 @@
-class Api::V1::BaseController < ApplicationController
+class Api::V1::BaseController < ActionController::API
```
So now the api controller and regular controllers do not have single parent ? //cc @vipulnsward @prathamesh-sonpatki 

You have moved common stuff like "setting honeybadger" to a module and that module is being included in both the controllers. However now `BasicUserSetting` has "layout carrier" which is not needed by api controllers. Hence I prefer having common stuff at "ApplicationController" rather than into a module.

prathamesh-sonpatki replied:
We can have same parent `ApplicationController`. API only apps have `ActionController::API` as their parent class. But as we are already inside normal Rails app, we can use `ApplicationController`.

neerajdotname replied:
@kalamkar In that case let's use common parent strategy.

neerajdotname commented:
- [ ] Not strictly a Rails 5 upgrade but lets also remove gem `browser`.  It would also mean remove `desktop` and `phone` folder from assets. I had thought they would be widely used but variants are rarely used.

neerajdotname commented:
https://monosnap.com/file/sYi5RnKdGcDhipTj3GQ5CXXhcQeQLH.png

kalamkar commented:
@neerajdotname _a Incorporated the suggested changes. Please review.

neerajdotname commented on Gemfile:
```diff
@@ -17,15 +17,14 @@ gem 'pg'
 
 # web interface for delayed job
+gem "sinatra", github: 'sinatra/sinatra'
+gem "rack-protection", github: 'sinatra/rack-protection'
```
We have comment before every gem to explain why it is needed. Please do the same for this gem.

neerajdotname commented on config/initializers/delayed_job_config.rb:
```diff
@@ -1,3 +1,5 @@
+MAIL_DELIVERY_PRIORITY_MEDIUM = -1
```
Please add comment regarding what does -1 do.

kalamkar commented:
@neerajdotname Please review.
